### DISCUSSION TITLE
OpenID authentication with Twitch

### DISCUSSION THREAD

**Discussion by y-ahlers** [directus/directus#9671] [Archive]
Hi guys,

I'm trying to set up login with twitch as auth provider and I'm getting the following error log message:

```
20:41:44 ⚠️  Service returned unexpected response
20:41:44 ✨ request completed GET 302 /auth/login/twitch/callback?code=<code>&scope=openid+user%3Aread%3Aemail
&state=<state> 2.4s
20:41:44 ✨ request completed GET 200 /admin/login?reason=SERVICE_UNAVAILABLE 1ms
```

So my first guess was that I maybe misconfigured some ENV variables for which reason I added a second configuration (like the docs with google) and that worked perfectly fine.

I went on and dove deeper into your code and recognized you are now using `openid-client` and that my configuration must end up in this line [api/src/auth/drivers/openid.ts#L268](https://github.com/directus/directus/blob/ca19cb1a999a030ee69703eda26dea809014665b/api/src/auth/drivers/openid.ts#L268). Consequently the `/callback` must be triggered but the `userinfo` is lacking of the email attribute. My guess is the `authenticationService.login()` throws because of that

Therefore I made a little sample repo and tried out myself to check what twitch is broadcasting back. The `TokenSet.claims()` only contains the attributes `email` and `email_verified` if I send the [claims](https://dev.twitch.tv/docs/authentication/getting-tokens-oidc#claims) to the authorize endpoint upfront, as you can see in my [project](https://github.com/y-ahlers/openid-twitch-example/blob/faebec96d80e951bea0edc00a6f12f3c44d370d8/index.js#L35).

Finally, my question, is there a way to configure these claims via some undocumented ENV variables or am I missing something else? Can you please help me out how I have to set up twitch as an auth provider?

Thank you very much!

**Comment by joselcvarela** [comment]
Hello @y-ahlers 
This was a though one to found.
The actual problem comes from Twitch itself and maybe openid-client.

When the request `https://id.twitch.tv/oauth2/token` is made to Twitch, in the response we got this:
```
{
  ...
  "token_type": "bearer"
}
```

Then the `bearer` is reused as `Authorization` header to request this `https://id.twitch.tv/oauth2/userinfo`. So we end up with an header like `Authorization: bearer ...`.
The problem is, Twitch not accepts `bearer` as lower case but needs to be capitalised instead. So, on one hand this `Authorization: bearer ...` returns 401, on the other hand this `Authorization: Bearer ...` returns 200.
Not sure who has fault here, Twitch or openid-client, but you can try open an issue on both.

**Reply by aidenfoxx** [reply]
This is fixed here: https://github.com/directus/directus/pull/9598 

I think the fault is on both, personally. "openid-client" expects the "token_type" to be valid for the service that sends it, which I would probably also expect, but they probably should have verified things a bit better! 😛

**Reply by joselcvarela** [reply]
Thank you @aidenfoxx 
If you got earlier you have spared me a lot of time 🥲

**Comment by aidenfoxx** [comment]
If a provider doesn't return the email address you can use a different key as the identifier (username, for example). The config for that is `AUTH_<provider>_IDENTIFIER_KEY`, documented here:

https://docs.directus.io/configuration/config-options/#openid

**Comment by joselcvarela** [comment]
Just for reference, this was my config for Twitch provider:
```.env
AUTH_PROVIDERS="twitch"

AUTH_TWITCH_DRIVER="openid"
AUTH_TWITCH_SCOPE="openid user:read:email"
AUTH_TWITCH_CLIENT_ID="..."
AUTH_TWITCH_CLIENT_SECRET="..."
AUTH_TWITCH_ISSUER_URL="https://id.twitch.tv/oauth2"
AUTH_TWITCH_IDENTIFIER_KEY="email"
AUTH_TWITCH_ICON="twitch"

**Reply by y-ahlers** [reply]
This is exactly what I have configured

**Comment by y-ahlers** [comment]
First of all, thank you for your replies!

I can confirm the TokenSet  of `openid-client` contains an all lower case "bearer", I get that too

TokenSet
```json
{
  "access_token": "<access_token>",
  "expires_at": 1636584234,
  "id_token": "<id_token>",
  "scope": ["openid","user:read:email"],
  "token_type": "bearer"
}
```

When I send the auth header with an all lower case bearer I also get ` {"status": 401,  "message": "missing authorization token"}`

As @aidenfoxx mentioned in the facebook-fix the `userinfo` will be fetched with the `access_token` if there is an `userinfo_endpoint`. This is what I get for this case

```json
 {
  "aud": "cy7jvfta9...",
  "exp": 1636570804,
  "iat": 1636569904,
  "iss": "https://id.twitch.tv/oauth2",
  "sub": "<sub> ",
  "azp": "cy7jvfta9c9n6sr...",
  "preferred_username": "<my_username>"
}
```

So if there is no `userinfo_endpoint` the `userinfo`-object will be the result of `tokenSet.claims()` , but the claims output will only contain the `email`  or the `email_verified` attributes until you call the authorization endpoint with:

```ts
const url = client.authorizationUrl({
    //...
    claims: {
      id_token: { email: null, email_verified: null },
    },
  });
``` 

In the first case when there is a `userinfo_endpoint` (for twitch there will be one) the `userinfo` object will never have the `email` because of the reason above. The `email` will only be in the claims output.

With the `grant` flow couple weeks ago I could easily set `AUTH_TWITCH_EMAIL_KEY="data[0].email"` but now no matter what I set in `AUTH_TWITCH_IDENTIFIER_KEY` I get the `SERVICE_UNAVAILABLE` error.

I now know twitch has a special/weird implementation of the OIDC auth code flow, but even if you successfully request the `userinfo_endpoint`  you have to set the claims, or am I wrong? Maybe I don't see the whole picture.

I hope this is not too confusing and I don't mess things up.

Thank you again!

**Reply by aidenfoxx** [reply]
If you get `SERVICE_UNAVAILABLE` there are bigger issues. As mentioned, things should work with #9598 which enforces "Bearer" with the token.

As for the email not being included in the profile response, it looks like twitch might require explicit "claims" in the authorize request to return that. That kind of thing would become possible if this feature is implemented https://github.com/directus/directus/discussions/9691.
